code/general/return-or-set-vs-or-equals.rb

```diff
@@ -0,0 +1,48 @@
+require 'benchmark/ips'
+
+# note: the more TIMES the memoized value is returned, the more profitable
+# it is to explicitly return the value without invoking the ||= operator.
+#
+# with TIMES=1: ||= is 1.07 times slower than early return
+# with TIMES=10: ||= is 1.20 times slower
+# with TIMES=100: ||= is 1.30 times slower
+# with TIMES=1000: ||= is 1.32 times slower
+# with TIMES=10000: ||= is 1.31 times slower
+
+TIMES = Integer(ARGV.fetch(0, 100))
+VALUE = 'some value'.freeze
+
+class Memoizer
+  def initialize
+    @value = nil
+  end
+
+  def value1
+    @value ||= VALUE
+  end
+
+  def value2
+    return @value if @value
```

Comment:
`return @value if defined? @value` is a real world use case for avoiding undefined exception.

Reply:
In my case I can ensure that the instance variable is always defined.

If I change the test from `if @value` to `if defined?(@value) && @value)` then the benchmark results are different and `||=` becomes the faster variant.

Reply:
`if defined?(@value) && @value)` is the equivalent of `||=`. So, maybe it makes sense for full comparison of all the use cases:

# For trying again if nil and not sure if variable is defined
1. `||=`
2. `return @value if defined?(@value) && @value)`

# For trying again if nil and sure the variable is defined
3. `return @value if @value`

# For not trying again if nil and not sure if variable is defined
4. `return @value if defined? @value`

# When sure that the variable is defined